src/p5xr/core/raycasting.js

```diff
@@ -0,0 +1,127 @@
+  let ray = {
+    origin: new p5.Vector(0, 0, 0),
+    direction: new p5.Vector()
+  };
+
+  let poseMatrix = p5xr.instance.viewer.poseMatrix.copy();
+  poseMatrix.transpose(poseMatrix);
+  poseMatrix = poseMatrix.mat4;
+
+  // set origin of ray to pose position
+  ray.origin.x = poseMatrix[3];
+  ray.origin.y = poseMatrix[7];
+  ray.origin.z = poseMatrix[11];
+  
+  let initialMVMatrix = p5xr.instance.viewer.initialMVMatrix.copy();
+  initialMVMatrix.transpose(initialMVMatrix);
+  initialMVMatrix = initialMVMatrix.mat4;
+
+  
+  // transform ray origin to view space
+  let rayOriginCopy = ray.origin.copy();
+  ray.origin.x = initialMVMatrix[0] * rayOriginCopy.x + initialMVMatrix[1] * rayOriginCopy.y + initialMVMatrix[2] * rayOriginCopy.z + initialMVMatrix[3];
+  ray.origin.y = initialMVMatrix[4] * rayOriginCopy.x + initialMVMatrix[5] * rayOriginCopy.y + initialMVMatrix[6] * rayOriginCopy.z + initialMVMatrix[7];
+  ray.origin.z = initialMVMatrix[8] * rayOriginCopy.x + initialMVMatrix[9] * rayOriginCopy.y + initialMVMatrix[10] * rayOriginCopy.z + initialMVMatrix[11];
+
+  // get ray direction from left eye
+  let leftDirection = new p5.Vector(screenX, screenY, -1);
+
+  let leftPMatrixInverse = new p5.Matrix();
+  leftPMatrixInverse.invert(p5xr.instance.viewer.leftPMatrix.copy());
+  leftPMatrixInverse.transpose(leftPMatrixInverse);
+  leftPMatrixInverse = leftPMatrixInverse.mat4;
+
+  let leftDirectionCopy = leftDirection.copy();
+  leftDirection.x = leftPMatrixInverse[0] * leftDirectionCopy.x + leftPMatrixInverse[1] * leftDirectionCopy.y + leftPMatrixInverse[2] * leftDirectionCopy.z + leftPMatrixInverse[3];
+  leftDirection.y = leftPMatrixInverse[4] * leftDirectionCopy.x + leftPMatrixInverse[5] * leftDirectionCopy.y + leftPMatrixInverse[6] * leftDirectionCopy.z + leftPMatrixInverse[7];
+  leftDirection.normalize();
+
+  // get ray direction from right eye
+  let rightDirection = new p5.Vector(screenX, screenY, -1);
+  
+  let rightPMatrixInverse = new p5.Matrix();
+  rightPMatrixInverse.invert(p5xr.instance.viewer.rightPMatrix.copy());
+  rightPMatrixInverse.transpose(rightPMatrixInverse);
+  rightPMatrixInverse = rightPMatrixInverse.mat4;
+
+  let rightDirectionCopy = rightDirection.copy();
+  rightDirection.x = rightPMatrixInverse[0] * rightDirectionCopy.x + rightPMatrixInverse[1] * rightDirectionCopy.y + rightPMatrixInverse[2] * rightDirectionCopy.z + rightPMatrixInverse[3];
+  rightDirection.y = rightPMatrixInverse[4] * rightDirectionCopy.x + rightPMatrixInverse[5] * rightDirectionCopy.y + rightPMatrixInverse[6] * rightDirectionCopy.z + rightPMatrixInverse[7];
+  rightDirection.normalize();
+
+  // combine both ray directions
+  ray.direction = p5.Vector.add(leftDirection, rightDirection).normalize();
+
+  return ray;
+}
+
+p5.prototype.intersectsSphere = function(radius, arg2, arg3) {
+  let ray;
+  if(arg3 !== undefined) {
+    ray = getRayFromScreen(arg2, arg3);
+  }
+  else {
+    ray = arg2;
+  }
+
+  if(ray === null)
```

Comment:
I think that in the event that a ray is not passed we should default to 0, 0. We are primarily envisioning this usage in VR right?

Reply:
Yeah. I have made it default to 0, 0.

Reply:
I realise now that defaulting screenX screenY will not be possible in `intersectsBox()`.
Here is the syntax : 
```
intersectsBox(Number width, [Number height], [Number depth], screenX, screenY)
intersectsBox(Number width, [Number height], [Number depth], rayObject ray)
- returns Boolean
```

If I default screenX and screenY, then a call like `intersectsBox(2, 0.5, 0.7)` would be ambiguous as the last 2 arguments could be depth, height or screenX, screenY.

Reply:
For organization purposes, I moved this discussion to the [intersectsBox PR](https://github.com/stalgiag/p5.xr/pull/49/files#r306069206)